oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/Optimizer.kt

```diff
@@ -0,0 +1,358 @@
+            val sender: String,
+            val message: String
+    )
+
+    data class Result(
+            val name: String,
+            val resultType: String,
+            val inputs: String,
+            val outputs: String
+    )
+
+    data class Simulation(
+            val inputs: List<Input>,
+            val outputs: List<Output>,
+            val description: String,
+            val input: StreamObserver<OASISQueryDTO>,
+            val output: Channel<SimulationResponseDTO>,
+            val update: Channel<NodeStatusCommandOrResponseDTO>,
+            val error: Channel<ErrorResponseDTO>,
+            val timeOut: Duration? = null
+    )
+
+    override fun changeNodeName(request: NodeNameChangeCommandDTO, responseObserver: StreamObserver<NodeNameChangeResponseDTO>) = responseObserver.consume {
+        val newNameSimulation = simulationsByName[request.newName]
+        val target = simulationsByName[request.oldName]
+        if (newNameSimulation == null && target != null) {
+            simulationsByName += request.newName to target
+            simulationsByName -= request.oldName
+
+            viewData.removeNode(request.oldName)
+            viewData.addNodes(request.newName)
+
+            NodeNameChangeResponseDTO.newBuilder().setChanged(true).build()
+        } else {
+            NodeNameChangeResponseDTO.newBuilder().setChanged(false).build()
+        }
+    }
+
+    override fun register(request: RegistrationCommandDTO, responseObserver: StreamObserver<OASISQueryDTO>) {
+        if (request.name in simulationsByName.keys) {
+            return
+        }
+        viewData.addNodes(request.name)
+        simulationsByName += request.name to Simulation(emptyList(), emptyList(), "", responseObserver, Channel(1), Channel(1), Channel(1))
+        viewData.updateStatusMessage("${request.name} registered")
+    }
+
+    fun unregisterAll() {
+        viewData.removeAll()
+        simulationsByName.values.forEach { sim ->
+            sim.input.onCompleted()
+            sim.output.close()
+        }
+        simulationsByName = emptyMap()
+    }
+
+    override fun startOptimization(request: StartOptimizationCommandDTO, responseObserver: StreamObserver<StartOptimizationResponseDTO>) = responseObserver.consume {
+        startOptimization(RandomNumberOptimizer())
+        StartOptimizationResponseDTO.newBuilder().setMessage("Started").setStarted(true).build()
+    }
+
+    sealed class SimResult(open val name: String, open val result: Map<String, Double>) {
+        data class Success(override val name: String, override val result: Map<String, Double>) : SimResult(name, result)
+        data class TimeOut(override val name: String) : SimResult(name, emptyMap())
+        data class TimeOutFailure(override val name: String, val exception: String) : SimResult(name, emptyMap())
+        data class Failure(override val name: String, val exception: String) : SimResult(name, emptyMap())
+    }
```

Comment:
this is a very strange combination of `open val` and `override val`s. 

on any one object record, at the very elast, you are going to have each piece of duplicated. CanI ask why you need `override val` instead of just using parameters, and leaving the `val name: String` left as final (read: not-`open`)?

Reply:
I think I wasn't doing any select on these result when I do the logging. Not exactly sure but I don't think I need those anymore